flax/core/lift.py

```diff
@@ -715,21 +723,39 @@ def jit(fn: Callable[..., Any],
     static_argnums = (static_argnums,)
   if not isinstance(donate_argnums, Iterable):
     donate_argnums = (donate_argnums,)
-  # offset argnums by one because first argument is the scope.
-  static_argnums = tuple(i + 1 for i in static_argnums if i > 0)
-  donate_argnums = tuple(i + 1 for i in donate_argnums if i > 0)
-  def inner(scope_fn, repack_fn, variable_groups, rng_groups, *args):
-    @functools.partial(jax.jit,
-                       static_argnums=static_argnums,
-                       donate_argnums=donate_argnums,
-                       device=device, backend=backend)
-    @functools.wraps(fn)
-    def jitted(variable_groups, rng_groups, *args):
-      scope = scope_fn(variable_groups, rng_groups)
-      y = fn(scope, *args)
-      return y, repack_fn(scope)
+  # offset argnums by two because first argument in the original function is the scope
+  # while jitted has 3 functions before the user arguments.
+  static_argnums = (0,) + tuple(i + 2 for i in static_argnums if i > 0)
+  donate_argnums = tuple(i + 2 for i in donate_argnums if i > 0)
+  
+  # Close over scope_fn & repack_fn to avoid recompilation
+  # this is impure but we use the fingerprint arg to differentiate between cases
+  # where scope_fn or repack_fn actually produce non-identical results.
+  scope_fn = None  # type: Callable
+  repack_fn = None  # type: Callable
+  @functools.partial(jax.jit,
+                     static_argnums=static_argnums,
+                     donate_argnums=donate_argnums,
+                     device=device, backend=backend)
+  @functools.wraps(fn)
+  def jitted(fingerprint, variable_groups, rng_groups, *args):
+    # fingerprint is only used to differentiate the cache signature for cases
```

Comment:
would it be useful to mark scope_fn, repack_fn 'nonlocal' here as well just to emphasize what we're doing for future readers?

Reply:
> A clever hack! what exactly does your note in the comment above mean? is this unsound when nested inside other transforms?

Just like jit the cache doesn't trigger if you call lift.jit multiple times which is what transforms.py ends up doing.